File: services/users/src/controllers/webhooks/lemonSqueezyWebhook.js
import User from '../../models/User.js';
import Subscription, { SUBSCRIPTION_PLANS } from '../../models/Subscription.js';
import { verifyWebhookSignature } from '../../utils/lemonsqueezy.js';
import axios from 'axios';
import { config } from '../../config/config.js';

// Handle Lemon Squeezy webhooks
export const handleLemonSqueezyWebhook = async (req, res) => {
  const startTime = Date.now();
  const requestId = `webhook_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  
  try {
    const signature = req.headers['x-signature'];
    const payload = JSON.stringify(req.body);
    
    console.log(`🍋 [${requestId}] Webhook received at ${new Date().toISOString()}`);
    console.log(`🍋 [${requestId}] Headers:`, {
      'x-signature': signature ? 'present' : 'missing',
      'content-type': req.headers['content-type'],
      'user-agent': req.headers['user-agent']
    });
    
    // Verify webhook signature
    if (!verifyWebhookSignature(payload, signature, process.env.LEMONSQUEEZY_WEBHOOK_SECRET)) {
      console.error(`🍋 [${requestId}] ❌ Invalid webhook signature`);
      return res.status(401).json({
        success: false,
        message: 'Invalid webhook signature',
        requestId
      });
    }

    const { meta, data } = req.body;
    const eventName = meta.event_name;

    console.log(`🍋 [${requestId}] ✅ Webhook signature verified`);
    console.log(`🍋 [${requestId}] Event: ${eventName}`);
    console.log(`🍋 [${requestId}] Data ID: ${data?.id}`);
    console.log(`🍋 [${requestId}] Custom Data:`, meta.custom_data || data?.attributes?.custom_data);

    switch (eventName) {
      case 'subscription_created':
        await handleSubscriptionCreated(data);
        break;
      
      case 'subscription_updated':
        await handleSubscriptionUpdated(data);
        break;
      
      case 'subscription_cancelled':
        await handleSubscriptionCancelled(data);
        break;
      
      case 'subscription_resumed':
        await handleSubscriptionResumed(data);
        break;
      
      case 'subscription_expired':
        await handleSubscriptionExpired(data);
        break;
      
      case 'subscription_paused':
        await handleSubscriptionPaused(data);
        break;
      
      case 'subscription_unpaused':
        await handleSubscriptionUnpaused(data);
        break;
      
      case 'subscription_payment_failed':
        await handleSubscriptionPaymentFailed(data);
        break;
      
      case 'subscription_payment_success':
        await handleSubscriptionPaymentSuccess(data);
        break;
      
      case 'order_created':
        await handleOrderCreated(data);
        break;
      
      default:
        console.log(`🍋 [${requestId}] ⚠️ Unhandled webhook event: ${eventName}`);
    }

    const processingTime = Date.now() - startTime;
    console.log(`🍋 [${requestId}] ✅ Webhook processed successfully in ${processingTime}ms`);
    
    res.status(200).json({ 
      success: true, 
      requestId,
      processingTime: `${processingTime}ms`
    });
  } catch (error) {
    const processingTime = Date.now() - startTime;
    console.error(`🍋 [${requestId}] ❌ Webhook processing error (${processingTime}ms):`, error);
    console.error(`🍋 [${requestId}] Stack trace:`, error.stack);
    
    res.status(500).json({
      success: false,
      message: 'Webhook processing failed',
      error: error.message,
      requestId,
      processingTime: `${processingTime}ms`
    });
  }
};

// Handle subscription created
const handleSubscriptionCreated = async (data) => {
  const subscriptionId = data.id;
  console.log(`🍋 [SUBSCRIPTION_CREATED] Processing subscription: ${subscriptionId}`);
  
  try {
    const customUserId = data.attributes.custom_data?.user_id;
    console.log(`🍋 [SUBSCRIPTION_CREATED] Custom user ID: ${customUserId}`);
    console.log(`🍋 [SUBSCRIPTION_CREATED] Variant ID: ${data.attributes.variant_id}`);
    console.log(`🍋 [SUBSCRIPTION_CREATED] Customer ID: ${data.attributes.customer_id}`);
    
    if (!customUserId) {
      console.error(`🍋 [SUBSCRIPTION_CREATED] ❌ No user_id in subscription created webhook for subscription ${subscriptionId}`);
      return;
    }

    console.log(`🍋 [SUBSCRIPTION_CREATED] Looking up user: ${customUserId}`);
    const user = await User.findById(customUserId).populate('subscriptionId');
    if (!user) {
      console.error(`🍋 [SUBSCRIPTION_CREATED] ❌ User not found: ${customUserId}`);
      return;
    }

    console.log(`🍋 [SUBSCRIPTION_CREATED] ✅ User found: ${user.email}`);
    const subscription = user.subscriptionId;
    if (!subscription) {
      console.error(`🍋 [SUBSCRIPTION_CREATED] ❌ Subscription not found for user: ${customUserId}`);
      return;
    }

    console.log(`🍋 [SUBSCRIPTION_CREATED] ✅ Subscription found: ${subscription._id}`);

    // Map variant ID to plan
    const variantToPlan = {
      [process.env.LEMONSQUEEZY_STANDARD_VARIANT_ID]: SUBSCRIPTION_PLANS.STANDARD,
      [process.env.LEMONSQUEEZY_PRO_VARIANT_ID]: SUBSCRIPTION_PLANS.PRO
    };

    const plan = variantToPlan[data.attributes.variant_id];
    if (!plan) {
      console.error(`🍋 [SUBSCRIPTION_CREATED] ❌ Unknown variant ID: ${data.attributes.variant_id}`);
      console.error(`🍋 [SUBSCRIPTION_CREATED] Available variants:`, Object.keys(variantToPlan));
      return;
    }

    console.log(`🍋 [SUBSCRIPTION_CREATED] ✅ Plan mapped: ${plan}`);
    console.log(`🍋 [SUBSCRIPTION_CREATED] Updating subscription in database...`);

    // Store old values for comparison
    const oldPlan = subscription.plan;
    const oldStatus = subscription.status;

    // Update subscription
    subscription.plan = plan;
    subscription.status = 'active';
    subscription.lemonSqueezySubscriptionId = data.id;
    subscription.lemonSqueezyCustomerId = data.attributes.customer_id;
    subscription.lemonSqueezyVariantId = data.attributes.variant_id;
    subscription.currentPeriodStart = new Date(data.attributes.created_at);
    subscription.currentPeriodEnd = new Date(data.attributes.renews_at);
    subscription.cancelAtPeriodEnd = false;
    subscription.cancelledAt = null;

    await subscription.save();
    
    console.log(`🍋 [SUBSCRIPTION_CREATED] ✅ Database updated:`);
    console.log(`🍋 [SUBSCRIPTION_CREATED]   Plan: ${oldPlan} → ${plan}`);
    console.log(`🍋 [SUBSCRIPTION_CREATED]   Status: ${oldStatus} → active`);
    console.log(`🍋 [SUBSCRIPTION_CREATED]   Period: ${data.attributes.created_at} → ${data.attributes.renews_at}`);
    
    // Invalidate gateway cache for this user
    console.log(`🍋 [SUBSCRIPTION_CREATED] Invalidating gateway cache...`);
    await invalidateGatewayCache(customUserId);
    
    console.log(`🍋 [SUBSCRIPTION_CREATED] ✅ Subscription successfully created for user ${user.email} (${customUserId}): ${plan}`);
  } catch (error) {
    console.error(`🍋 [SUBSCRIPTION_CREATED] ❌ Error handling subscription created:`, error);
    console.error(`🍋 [SUBSCRIPTION_CREATED] Stack trace:`, error.stack);
  }
};

// Handle subscription updated
const handleSubscriptionUpdated = async (data) => {
  try {
    const subscription = await Subscription.findOne({
      lemonSqueezySubscriptionId: data.id
    });

    if (!subscription) {
      console.error(`Subscription not found: ${data.id}`);
      return;
    }

    // Update subscription details
    subscription.status = data.attributes.status;
    subscription.currentPeriodStart = new Date(data.attributes.created_at);
    subscription.currentPeriodEnd = new Date(data.attributes.renews_at);

    await subscription.save();
    
    // Invalidate gateway cache for this user
    const user = await User.findOne({ subscriptionId: subscription._id });
    if (user) {
      await invalidateGatewayCache(user._id.toString());
    }
    
    console.log(`Subscription updated: ${data.id}`);
  } catch (error) {
    console.error('Error handling subscription updated:', error);
  }
};

// Handle subscription cancelled
const handleSubscriptionCancelled = async (data) => {
  try {
    const subscription = await Subscription.findOne({
      lemonSqueezySubscriptionId: data.id
    });

    if (!subscription) {
      console.error(`Subscription not found: ${data.id}`);
      return;
    }

    subscription.status = 'cancelled';
    subscription.cancelAtPeriodEnd = true;
    subscription.cancelledAt = new Date(data.attributes.cancelled_at);

    await subscription.save();
    
    // Invalidate gateway cache for this user
    const user = await User.findOne({ subscriptionId: subscription._id });
    if (user) {
      await invalidateGatewayCache(user._id.toString());
    }
    
    console.log(`Subscription cancelled: ${data.id}`);
  } catch (error) {
    console.error('Error handling subscription cancelled:', error);
  }
};

// Handle subscription resumed
const handleSubscriptionResumed = async (data) => {
  try {
    const subscription = await Subscription.findOne({
      lemonSqueezySubscriptionId: data.id
    });

    if (!subscription) {
      console.error(`Subscription not found: ${data.id}`);
      return;
    }

    subscription.status = 'active';
    subscription.cancelAtPeriodEnd = false;
    subscription.cancelledAt = null;

    await subscription.save();
    
    // Invalidate gateway cache for this user
    const user = await User.findOne({ subscriptionId: subscription._id });
    if (user) {
      await invalidateGatewayCache(user._id.toString());
    }
    
    console.log(`Subscription resumed: ${data.id}`);
  } catch (error) {
    console.error('Error handling subscription resumed:', error);
  }
};

// Handle subscription expired
const handleSubscriptionExpired = async (data) => {
  try {
    const subscription = await Subscription.findOne({
      lemonSqueezySubscriptionId: data.id
    });

    if (!subscription) {
      console.error(`Subscription not found: ${data.id}`);
      return;
    }

    // Downgrade to free plan
    subscription.plan = SUBSCRIPTION_PLANS.FREE;
    subscription.status = 'expired';
    subscription.lemonSqueezySubscriptionId = null;
    subscription.lemonSqueezyVariantId = null;
    subscription.currentPeriodEnd = null;

    await subscription.save();
    
    // Invalidate gateway cache for this user
    const user = await User.findOne({ subscriptionId: subscription._id });
    if (user) {
      await invalidateGatewayCache(user._id.toString());
    }
    
    console.log(`Subscription expired, downgraded to free: ${data.id}`);
  } catch (error) {
    console.error('Error handling subscription expired:', error);
  }
};

// Handle subscription paused
const handleSubscriptionPaused = async (data) => {
  try {
    const subscription = await Subscription.findOne({
      lemonSqueezySubscriptionId: data.id
    });

    if (!subscription) {
      console.error(`Subscription not found: ${data.id}`);
      return;
    }

    subscription.status = 'paused';
    await subscription.save();
    
    // Invalidate gateway cache for this user
    const user = await User.findOne({ subscriptionId: subscription._id });
    if (user) {
      await invalidateGatewayCache(user._id.toString());
    }
    
    console.log(`Subscription paused: ${data.id}`);
  } catch (error) {
    console.error('Error handling subscription paused:', error);
  }
};

// Handle subscription unpaused
const handleSubscriptionUnpaused = async (data) => {
  try {
    const subscription = await Subscription.findOne({
      lemonSqueezySubscriptionId: data.id
    });

    if (!subscription) {
      console.error(`Subscription not found: ${data.id}`);
      return;
    }

    subscription.status = 'active';
    await subscription.save();
    
    // Invalidate gateway cache for this user
    const user = await User.findOne({ subscriptionId: subscription._id });
    if (user) {
      await invalidateGatewayCache(user._id.toString());
    }
    
    console.log(`Subscription unpaused: ${data.id}`);
  } catch (error) {
    console.error('Error handling subscription unpaused:', error);
  }
};

// Handle subscription payment failed
const handleSubscriptionPaymentFailed = async (data) => {
  try {
    const subscription = await Subscription.findOne({
      lemonSqueezySubscriptionId: data.id
    });

    if (!subscription) {
      console.error(`Subscription not found: ${data.id}`);
      return;
    }

    subscription.status = 'past_due';
    await subscription.save();
    
    // Invalidate gateway cache for this user
    const user = await User.findOne({ subscriptionId: subscription._id });
    if (user) {
      await invalidateGatewayCache(user._id.toString());
    }
    
    console.log(`Subscription payment failed: ${data.id}`);
  } catch (error) {
    console.error('Error handling subscription payment failed:', error);
  }
};

// Handle subscription payment success
const handleSubscriptionPaymentSuccess = async (data) => {
  try {
    const subscription = await Subscription.findOne({
      lemonSqueezySubscriptionId: data.id
    });

    if (!subscription) {
      console.error(`Subscription not found: ${data.id}`);
      return;
    }

    subscription.status = 'active';
    
    // Add invoice to history
    subscription.invoices.push({
      lemonSqueezyInvoiceId: data.attributes.invoice_id,
      amount: data.attributes.total,
      currency: data.attributes.currency,
      status: 'paid',
      paidAt: new Date()
    });

    await subscription.save();
    
    // Invalidate gateway cache for this user
    const user = await User.findOne({ subscriptionId: subscription._id });
    if (user) {
      await invalidateGatewayCache(user._id.toString());
    }
    
    console.log(`Subscription payment successful: ${data.id}`);
  } catch (error) {
    console.error('Error handling subscription payment success:', error);
  }
};

// Handle order created (one-time purchases)
const handleOrderCreated = async (data) => {
  try {
    console.log(`Order created: ${data.id}`);
    // Handle one-time purchases if needed
  } catch (error) {
    console.error('Error handling order created:', error);
  }
};

// Helper function to invalidate gateway cache
const invalidateGatewayCache = async (userId) => {
  try {
    const gatewayUrl = process.env.GATEWAY_URL || 'http://localhost:8080';
    
    // Call gateway cache invalidation endpoint
    await axios.delete(`${gatewayUrl}/api/v1/cache/user/${userId}`, {
      headers: {
        'X-API-Key': config.globalApiKey
      },
      timeout: 5000 // 5 second timeout
    });
    
    console.log(`Gateway cache invalidated for user: ${userId}`);
  } catch (error) {
    // Don't fail the webhook if cache invalidation fails
    console.warn(`Failed to invalidate gateway cache for user ${userId}:`, error.message);
  }
};

File: services/common/services/lemonsqueezy_webhook.go
package services

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/seentics/seentics/services/common/config"
	"github.com/seentics/seentics/services/common/database"
	"github.com/sirupsen/logrus"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Subscription plans mapping
var SUBSCRIPTION_PLANS = map[string]string{
	"free":     "free",
	"standard": "standard",
	"pro":      "pro",
}

// Lemon Squeezy webhook data structure
type LemonSqueezyWebhookData struct {
	ID         string `json:"id"`
	Attributes struct {
		CustomData struct {
			UserID string `json:"user_id"`
		} `json:"custom_data"`
		VariantID   string    `json:"variant_id"`
		CustomerID  string    `json:"customer_id"`
		Status      string    `json:"status"`
		CreatedAt   time.Time `json:"created_at"`
		RenewsAt    time.Time `json:"renews_at"`
		CancelledAt time.Time `json:"cancelled_at"`
		InvoiceID   string    `json:"invoice_id"`
		Total       int       `json:"total"`
		Currency    string    `json:"currency"`
	} `json:"attributes"`
}

// Subscription model for database operations
type Subscription struct {
	ID                         primitive.ObjectID `bson:"_id,omitempty"`
	UserID                     primitive.ObjectID `bson:"userId"`
	Plan                       string             `bson:"plan"`
	Status                     string             `bson:"status"`
	LemonSqueezyCustomerID     string             `bson:"lemonSqueezyCustomerId"`
	LemonSqueezySubscriptionID string             `bson:"lemonSqueezySubscriptionId"`
	LemonSqueezyVariantID      string             `bson:"lemonSqueezyVariantId"`
	CurrentPeriodStart         time.Time          `bson:"currentPeriodStart"`
	CurrentPeriodEnd           *time.Time         `bson:"currentPeriodEnd"`
	CancelAtPeriodEnd          bool               `bson:"cancelAtPeriodEnd"`
	CancelledAt                *time.Time         `bson:"cancelledAt"`
	CreatedAt                  time.Time          `bson:"createdAt"`
	UpdatedAt                  time.Time          `bson:"updatedAt"`
}

// Webhook Services - Lemon Squeezy Integration
func HandleLemonSqueezyWebhook(c *gin.Context) {
	// Get webhook signature
	signature := c.GetHeader("X-Signature")
	if signature == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Missing webhook signature"})
		return
	}

	// Read raw body
	body, err := c.GetRawData()
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Failed to read request body"})
		return
	}

	// Verify webhook signature
	cfg := config.Load()
	if !verifyLemonSqueezySignature(string(body), signature, cfg.LemonSqueezyWebhookSecret) {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid webhook signature"})
		return
	}

	// Parse webhook payload
	var webhook struct {
		Meta struct {
			EventName string `json:"event_name"`
		} `json:"meta"`
		Data LemonSqueezyWebhookData `json:"data"`
	}

	if err := json.Unmarshal(body, &webhook); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid JSON payload"})
		return
	}

	eventName := webhook.Meta.EventName
	logrus.Info("Received Lemon Squeezy webhook:", eventName)

	// Handle different webhook events
	switch eventName {
	case "subscription_created":
		err = handleSubscriptionCreated(webhook.Data)
	case "subscription_updated":
		err = handleSubscriptionUpdated(webhook.Data)
	case "subscription_cancelled":
		err = handleSubscriptionCancelled(webhook.Data)
	case "subscription_resumed":
		err = handleSubscriptionResumed(webhook.Data)
	case "subscription_expired":
		err = handleSubscriptionExpired(webhook.Data)
	case "subscription_paused":
		err = handleSubscriptionPaused(webhook.Data)
	case "subscription_unpaused":
		err = handleSubscriptionUnpaused(webhook.Data)
	case "subscription_payment_failed":
		err = handleSubscriptionPaymentFailed(webhook.Data)
	case "subscription_payment_success":
		err = handleSubscriptionPaymentSuccess(webhook.Data)
	case "order_created":
		err = handleOrderCreated(webhook.Data)
	default:
		logrus.Warn("Unhandled webhook event:", eventName)
	}

	if err != nil {
		logrus.Error("Webhook processing error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Webhook processing failed"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}

// Verify Lemon Squeezy webhook signature
func verifyLemonSqueezySignature(payload, signature, secret string) bool {
	if secret == "" {
		logrus.Warn("Lemon Squeezy webhook secret not configured")
		return false
	}

	h := hmac.New(sha256.New, []byte(secret))
	h.Write([]byte(payload))
	expectedSignature := hex.EncodeToString(h.Sum(nil))

	return hmac.Equal([]byte(signature), []byte(expectedSignature))
}

// Handle subscription created
func handleSubscriptionCreated(data LemonSqueezyWebhookData) error {
	userID := data.Attributes.CustomData.UserID
	if userID == "" {
		logrus.Error("No user_id in subscription created webhook")
		return nil
	}

	userObjectID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		logrus.Error("Invalid user ID:", userID)
		return err
	}

	// Get variant to plan mapping from config
	cfg := config.Load()
	var plan string
	switch data.Attributes.VariantID {
	case cfg.LemonSqueezyStandardVariantID:
		plan = SUBSCRIPTION_PLANS["standard"]
	case cfg.LemonSqueezyProVariantID:
		plan = SUBSCRIPTION_PLANS["pro"]
	default:
		logrus.Error("Unknown variant ID:", data.Attributes.VariantID)
		return nil
	}

	// Find or create subscription
	collection := database.GetUserDB().Collection("subscriptions")

	subscription := Subscription{
		UserID:                     userObjectID,
		Plan:                       plan,
		Status:                     "active",
		LemonSqueezySubscriptionID: data.ID,
		LemonSqueezyCustomerID:     data.Attributes.CustomerID,
		LemonSqueezyVariantID:      data.Attributes.VariantID,
		CurrentPeriodStart:         data.Attributes.CreatedAt,
		CurrentPeriodEnd:           &data.Attributes.RenewsAt,
		CancelAtPeriodEnd:          false,
		CancelledAt:                nil,
		CreatedAt:                  time.Now(),
		UpdatedAt:                  time.Now(),
	}

	// Upsert subscription
	_, err = collection.ReplaceOne(
		context.Background(),
		bson.M{"userId": userObjectID},
		subscription,
	)
	if err != nil {
		logrus.Error("Failed to create/update subscription:", err)
		return err
	}

	// Update user subscription info
	userCollection := database.GetUserDB().Collection("users")
	_, err = userCollection.UpdateOne(
		context.Background(),
		bson.M{"_id": userObjectID},
		bson.M{"$set": bson.M{
			"subscriptionPlan":   plan,
			"subscriptionStatus": "active",
			"updatedAt":          time.Now(),
		}},
	)
	if err != nil {
		logrus.Error("Failed to update user subscription:", err)
		return err
	}

	logrus.Info("Subscription created for user:", userID, "plan:", plan)
	return nil
}

// Handle subscription updated
func handleSubscriptionUpdated(data LemonSqueezyWebhookData) error {
	collection := database.GetUserDB().Collection("subscriptions")

	update := bson.M{
		"$set": bson.M{
			"status":             data.Attributes.Status,
			"currentPeriodStart": data.Attributes.CreatedAt,
			"currentPeriodEnd":   data.Attributes.RenewsAt,
			"updatedAt":          time.Now(),
		},
	}

	_, err := collection.UpdateOne(
		context.Background(),
		bson.M{"lemonSqueezySubscriptionId": data.ID},
		update,
	)
	if err != nil {
		logrus.Error("Failed to update subscription:", err)
		return err
	}

	logrus.Info("Subscription updated:", data.ID)
	return nil
}

// Handle subscription cancelled
func handleSubscriptionCancelled(data LemonSqueezyWebhookData) error {
	collection := database.GetUserDB().Collection("subscriptions")

	update := bson.M{
		"$set": bson.M{
			"status":            "cancelled",
			"cancelAtPeriodEnd": true,
			"cancelledAt":       data.Attributes.CancelledAt,
			"updatedAt":         time.Now(),
		},
	}

	_, err := collection.UpdateOne(
		context.Background(),
		bson.M{"lemonSqueezySubscriptionId": data.ID},
		update,
	)
	if err != nil {
		logrus.Error("Failed to cancel subscription:", err)
		return err
	}

	logrus.Info("Subscription cancelled:", data.ID)
	return nil
}

// Handle subscription resumed
func handleSubscriptionResumed(data LemonSqueezyWebhookData) error {
	collection := database.GetUserDB().Collection("subscriptions")

	update := bson.M{
		"$set": bson.M{
			"status":            "active",
			"cancelAtPeriodEnd": false,
			"cancelledAt":       nil,
			"updatedAt":         time.Now(),
		},
	}

	_, err := collection.UpdateOne(
		context.Background(),
		bson.M{"lemonSqueezySubscriptionId": data.ID},
		update,
	)
	if err != nil {
		logrus.Error("Failed to resume subscription:", err)
		return err
	}

	logrus.Info("Subscription resumed:", data.ID)
	return nil
}

// Handle subscription expired
func handleSubscriptionExpired(data LemonSqueezyWebhookData) error {
	collection := database.GetUserDB().Collection("subscriptions")

	// Find subscription to get user ID
	var subscription Subscription
	err := collection.FindOne(
		context.Background(),
		bson.M{"lemonSqueezySubscriptionId": data.ID},
	).Decode(&subscription)
	if err != nil {
		logrus.Error("Subscription not found:", data.ID)
		return err
	}

	// Downgrade to free plan
	update := bson.M{
		"$set": bson.M{
			"plan":                       SUBSCRIPTION_PLANS["free"],
			"status":                     "expired",
			"lemonSqueezySubscriptionId": nil,
			"lemonSqueezyVariantId":      nil,
			"currentPeriodEnd":           nil,
			"updatedAt":                  time.Now(),
		},
	}

	_, err = collection.UpdateOne(
		context.Background(),
		bson.M{"lemonSqueezySubscriptionId": data.ID},
		update,
	)
	if err != nil {
		logrus.Error("Failed to expire subscription:", err)
		return err
	}

	// Update user subscription info
	userCollection := database.GetUserDB().Collection("users")
	_, err = userCollection.UpdateOne(
		context.Background(),
		bson.M{"_id": subscription.UserID},
		bson.M{"$set": bson.M{
			"subscriptionPlan":   SUBSCRIPTION_PLANS["free"],
			"subscriptionStatus": "expired",
			"updatedAt":          time.Now(),
		}},
	)
	if err != nil {
		logrus.Error("Failed to update user subscription:", err)
		return err
	}

	logrus.Info("Subscription expired, downgraded to free:", data.ID)
	return nil
}

// Handle subscription paused
func handleSubscriptionPaused(data LemonSqueezyWebhookData) error {
	collection := database.GetUserDB().Collection("subscriptions")

	update := bson.M{
		"$set": bson.M{
			"status":    "paused",
			"updatedAt": time.Now(),
		},
	}

	_, err := collection.UpdateOne(
		context.Background(),
		bson.M{"lemonSqueezySubscriptionId": data.ID},
		update,
	)
	if err != nil {
		logrus.Error("Failed to pause subscription:", err)
		return err
	}

	logrus.Info("Subscription paused:", data.ID)
	return nil
}

// Handle subscription unpaused
func handleSubscriptionUnpaused(data LemonSqueezyWebhookData) error {
	collection := database.GetUserDB().Collection("subscriptions")

	update := bson.M{
		"$set": bson.M{
			"status":    "active",
			"updatedAt": time.Now(),
		},
	}

	_, err := collection.UpdateOne(
		context.Background(),
		bson.M{"lemonSqueezySubscriptionId": data.ID},
		update,
	)
	if err != nil {
		logrus.Error("Failed to unpause subscription:", err)
		return err
	}

	logrus.Info("Subscription unpaused:", data.ID)
	return nil
}

// Handle subscription payment failed
func handleSubscriptionPaymentFailed(data LemonSqueezyWebhookData) error {
	collection := database.GetUserDB().Collection("subscriptions")

	update := bson.M{
		"$set": bson.M{
			"status":    "past_due",
			"updatedAt": time.Now(),
		},
	}

	_, err := collection.UpdateOne(
		context.Background(),
		bson.M{"lemonSqueezySubscriptionId": data.ID},
		update,
	)
	if err != nil {
		logrus.Error("Failed to update subscription payment failed:", err)
		return err
	}

	logrus.Info("Subscription payment failed:", data.ID)
	return nil
}

// Handle subscription payment success
func handleSubscriptionPaymentSuccess(data LemonSqueezyWebhookData) error {
	collection := database.GetUserDB().Collection("subscriptions")

	// Create invoice record
	invoice := bson.M{
		"lemonSqueezyInvoiceId": data.Attributes.InvoiceID,
		"amount":                data.Attributes.Total,
		"currency":              data.Attributes.Currency,
		"status":                "paid",
		"paidAt":                time.Now(),
		"createdAt":             time.Now(),
	}

	update := bson.M{
		"$set": bson.M{
			"status":    "active",
			"updatedAt": time.Now(),
		},
		"$push": bson.M{
			"invoices": invoice,
		},
	}

	_, err := collection.UpdateOne(
		context.Background(),
		bson.M{"lemonSqueezySubscriptionId": data.ID},
		update,
	)
	if err != nil {
		logrus.Error("Failed to update subscription payment success:", err)
		return err
	}

	logrus.Info("Subscription payment successful:", data.ID)
	return nil
}

// Handle order created (one-time purchases)
func handleOrderCreated(data LemonSqueezyWebhookData) error {
	logrus.Info("Order created:", data.ID)
	// Handle one-time purchases if needed
	return nil
}
